### ISSUE TITLE
Info on State Prefetching

### ISSUE THREAD

**Issue by jon-chuang** [ethereum/go-ethereum#22724]
Dear Geth Team,

I am currently looking into some optimisations to Geth related to parallelism, state fetching and snapshots as part of my work on the downstream `celo-blockchain`. 

Although it is not a priority, I think there is a chance that some work done could end up being merged upstream. Even if not, I think it is still a good idea to start a conversation.

I have a number of questions:
1. Is it correct to say as of now no actual optimisations exist based on the Berlin activated `AccessList` in terms of `Trie` (i.e. DB warming) or even snapshot object prefetching? What are the team’s plans for such things going forward? More generally, is there some sort of roadmap capturing planned features or are these only captured ephemerally in discussions/issues? 
2. Somewhat relatedly, is it correct to say that the `core/state_prefetcher.go` code is outdated in the sense that one would not like to call `TransitionDB` in advance to warm the state, as it is costly, what’s more, in parallel? What is the (future) purpose of this code since it’s currently not invoked anywhere as far as I can tell? Is the future for this code to morph into DB prefetching based on AccessList? If so, what would be a good entry point to spawning the prefetching? 
3. Why is it that the `TriePrefetcher` that lives in the `StateDB` uses one go worker per state root to walk the Trie? This seems inefficient to me, and in my mind, one should spawn as many goroutines as there are `trie.TryGet`s to be walked. In my mind, one ought to maintain a hashmap of `Tries` and pop tasks off a global task list and spawn as many goroutines as there are tasks. The global prefetcher loop will then respond to stop commands.. This way, one can leverage concurrency for `Trie` prefetching even within a single contract’s state Trie. Since `disk fetch >> goroutine context switch`, I think this is a reasonable change. Since this is a relatively isolated change, I guess I could submit a PR here to familiarise with the contribution process on geth. I guess I’ll make an issue first. One question is whether it is preferable to maintain copies of Tries per thread (since this is rather lightweight - just a pointer to a global? `Database` and the root node). Further, if the intention is to warm the underlying DB, there is not much reason (apart, perhaps, for convenience) to continue to maintain the `Trie`s in memory, is there? In essence, one can request for the `Trie`s anew from the `Database` from a separate location after deleting the corresponding prefetcher `Trie` object in memory, as long as one is maintaining a single global `Database` object.

Cheers,
Jon

**Comment by holiman** [comment]
> Is it correct to say as of now no actual optimisations exist based on the Berlin activated AccessList in terms of Trie (i.e. DB warming) or even snapshot object prefetching? What are the team’s plans for such things going forward? More generally, is there some sort of roadmap capturing planned features or are these only captured ephemerally in discussions/issues?

There are currently no such optimizations, correct. I also don't think anyone is working on adding such optimizations -- note, the access lists are not expected to be used by 'mainstream' users anytime soon. If that changes, and they start becoming more common, then it might be worth looking into properly. 

> Somewhat relatedly, is it correct to say that the core/state_prefetcher.go code is outdated

Yes, I'd say so. That prefetcher is only active when you are catchiing up, and have future blocks already downloaded. We've discussed removing it several times, it may just cause problems due to high IO. 

The `core/state/trie_prefetcher` is a better approach, where we resolve the trie based on what we're going to write. It works in all modes of operation, not just when catching up. Look into that for more details on how it works -- you can also read up some more info here: https://github.com/ethereum/go-ethereum/pull/21047 

> Why is it that the TriePrefetcher that lives in the StateDB uses one go worker per state root to walk the Trie? 

IIUC, you mean that we would load tries into one trie concurrently? AFAIK the trie is not safe to be used like this. The trie is concurrency-safe in the sense that any time it resolves something, it creates new nodes "on the way up" and you get a new root element. But your idea would be, as I understand it, to insert two things at the same time into the trie, which I don't think will work. 

I guess one could build a new trie structure. At one point, I PR:ed a 'batch-mode' for the trie, in which I just updated the nodes in place without replacing them. That _would_ make it possible to insert several things at once, as long as they operated on different subpaths in the trie. Maybe. It's not a trivial problem, feel free to see if you can sketch something up. 

> Further, if the intention is to warm the underlying DB, there is not much reason (apart, perhaps, for convenience) to continue to maintain the Tries in memory, is there?
 
No, the idea i snot to just warm the db, but to actually hand out a fully expanded trie to the caller.

**Comment by jon-chuang** [comment]
Hi @holiman , thank for the very detailed and quick reply!

Some replies and more questions:

1. Am I right to say that as of now, all `Trie`s one expects to encounter are in fact `SecureTrie`s? 

2.  > But your idea would be, as I understand it, to insert two things at the same time into the trie, which I don't think will work.

Actually, I just want to walk the `Trie`, i.e. prefetch concurrently (via `trie.TryGet`) within each `Trie`. I think insertion/mutation more generally is handled by a downstream consumer of the prefetcher?

TryGet mutates state iff `trie.root.(type) == HashNode`. However, this branch is defensive programming and is never invoked since `core/state/database.OpenTrie -> trie.NewSecure -> trie.New` sets `Trie.root` to a resolved node.

3. > No, the idea is not to just warm the db, but to actually hand out a fully expanded trie to the caller.

Ok, I agree here. `trie.Database` actually has a lot of state. Nonetheless, `Trie` and `SecureTrie` themselves are essentially stateless wrappers (apart from some cacheing and ...metrics - on number of leaf inserts, not relevant to our read-only use-case). Furthermore, in the comment description of `trie.Database`:
```
// Note, the trie Database is **not** thread safe in its mutations, but it **is**
// thread safe in providing individual, independent node access.
```
So calling `Trie.TryGet` concurrently, the only method called in `triePrefetcher` ought not to be a problem. Thus, I think maintaining copies of a Trie in each prefetcher goroutine, and then returning a copy of an arbitrary copy of the Trie referencing the underlying database ought not to be a problem. 

Alternatively, maintaining a reference in each goroutine worker to a given `Trie` living in a `map` that lives in the global `TriePrefetcher` and calling `TryGet` from it concurrently ought not to be a problem. This is the direction I prefer.

**Comment by jon-chuang** [comment]
So, in case I was a little too slow in understanding your comment, my main question now, @holiman is whether you think adding a mutex to `NodeFlag` would be acceptable. Possibly this could lead to concurrent `Trie` updates too. But unclear to me how useful this is.

My rationale for exploring the `Trie`-level prefetch concurrency is if txs are dominated by a single contract like an ERC-20. Since this is as far as I can tell not the situation for ETH TXs, it's unclear how much the bench that I linked above reflects such a workload.

**Comment by Jaysonmaldx** [comment]
Thank u
